beater/beater.go

```diff
@@ -1026,3 +1034,43 @@ func queryClusterUUID(ctx context.Context, esClient elasticsearch.Client) error
 	s.Set(response.ClusterUUID)
 	return nil
 }
+
+func adjustMaxProcs(ctx context.Context, d time.Duration, logger *logp.Logger) error {
+	setMaxProcs := func() {
+		if _, err := maxprocs.Set(maxprocs.Logger(logger.Infof)); err != nil {
+			logger.Errorf("failed to set GOMAXPROCS: %v", err)
+		}
+	}
+	// set the gomaxprocs immediately.
+	setMaxProcs()
+	ticker := time.NewTicker(d)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-ticker.C:
+			setMaxProcs()
+		}
+	}
+}
+
+type zapDiffCore struct {
```

Comment:
I'm not sure if we want to move this `zapDiffCore` struct to a different package like `log`

Reply:
IMO defer that until we want to reuse it. Hasn't come up before.

Reply:
I think we can simplify this though. e.g.

```go
var last string
log := func(format string, args ...interface{}) {
    s := fmt.Sprintf(format, args...)
    if s != last {
        log.Infof("%s", s)
        last = s
    }
}

...
```